lib/AST/LocalizationFormat.cpp

```diff
@@ -154,19 +169,25 @@ readYAML(llvm::yaml::IO &io, T &Seq, bool, Context &Ctx) {
     if (io.preflightElement(i, SaveInfo)) {
       DiagnosticNode current;
       yamlize(io, current, true, Ctx);
-      io.postflightElement(SaveInfo);
 
       // A diagnostic ID might be present in YAML and not in `.def` file,
       // if that's the case ScalarEnumerationTraits will assign the diagnostic ID
       // to `LocalDiagID::NumDiags`. Since the diagnostic ID isn't available
       // in `.def` it shouldn't be stored in the diagnostics array.
       if (current.id != LocalDiagID::NumDiags) {
+        io.postflightElement(SaveInfo);
+
         // YAML file isn't guaranteed to have diagnostics in order of their
         // declaration in `.def` files, to accommodate that we need to leave
         // holes in diagnostic array for diagnostics which haven't yet been
         // localized and for the ones that have `DiagnosticNode::id`
         // indicates their position.
         Seq[static_cast<unsigned>(current.id)] = std::move(current.msg);
+      } else {
+        DiagnosticID removedID;
```

Comment:
I don't understand this - I think it's okay to read DiagnosticNode and just ignore `msg` instead of having it read a separate type of node here.

Reply:
I don't think so, because the `DiagnosticNode` struct has an `id` of type `u_int32` and we need to store the diagnostic ID as a string. Therefore, I think we need a separate type of node which has an `id` as a string. WDYT?

Reply:
Could we make `id` of the DIagnosticNode a union instead so it would be `uint32_t` when matched and `std::string` otherwise?